File: bindings/sandbox.ts
import koffi from 'koffi';
import * as path from 'path';

// Define the C structs
export const SandboxMount = koffi.struct('sandbox_mount_t', {
    dir: 'const char*',
    read_only: 'int'
});

export const SandboxConfig = koffi.struct('sandbox_config_t', {
    mounts: koffi.pointer(SandboxMount),
    mount_count: 'size_t',
    disable_network: 'int',
    env_vars: koffi.pointer('const char*'),
    env_count: 'size_t',
    timeout_secs: 'unsigned int',
    stdout_buffer: koffi.pointer('char*'),
    stdout_size: koffi.pointer('size_t'),
    stderr_buffer: koffi.pointer('char*'),
    stderr_size: koffi.pointer('size_t'),
    max_memory_mb: 'unsigned int',
    max_cpu_percent: 'unsigned int',
    drop_privileges: 'int',
    target_uid: 'unsigned int',
    target_gid: 'unsigned int',
    denied_syscalls: koffi.pointer('const char*'),
    denied_syscall_count: 'size_t',
    seccomp_profile_path: 'const char*',
    apparmor_profile: 'const char*',
    use_pty: 'int',
    max_network_mbps: 'unsigned int'
});

const InitProto = koffi.proto('int InitProto(void)');
const ExecuteProto = koffi.proto('int ExecuteProto(sandbox_config_t*, int, const char**)');
const CleanupProto = koffi.proto('void CleanupProto(void)');

export const SandboxEngine = koffi.struct('sandbox_engine_t', {
    engine_name: 'const char*',
    description: 'const char*',
    init: koffi.pointer(InitProto),
    execute: koffi.pointer(ExecuteProto),
    execute_async: 'void*',
    wait_process: 'void*',
    free_process: 'void*',
    cleanup: koffi.pointer(CleanupProto)
});

export class LibMountSandbox {
    private lib: koffi.IKoffiLib;
    private getSandboxEngine: any;

    constructor(libPath: string = './build/libmountsandbox.so') {
        const ext = path.extname(libPath);
        let actualPath = libPath;
        if (!ext || ext === '.so' || ext === '.dylib' || ext === '.dll') {
            const platformExt = process.platform === 'darwin' ? '.dylib' : (process.platform === 'win32' ? '.dll' : '.so');
            actualPath = libPath.replace(/\.(so|dylib|dll)$/, platformExt);
        }
        this.lib = koffi.load(actualPath);
        this.getSandboxEngine = this.lib.func('int get_sandbox_engine(const char* name, _Out_ sandbox_engine_t** engine_out)');
    }

    public execute(engineName: string, commandArgs: string[], config: any = null): number {
        let enginePtrRef = [null];
        const res = this.getSandboxEngine(engineName, enginePtrRef);
        if (res !== 0 || !enginePtrRef[0]) {
            throw new Error(`Unknown sandbox engine: ${engineName}`);
        }
        const enginePtr = enginePtrRef[0];

        const engine = koffi.decode(enginePtr, SandboxEngine);

        const init = koffi.decode(engine.init, InitProto);
        const execute = koffi.decode(engine.execute, ExecuteProto);
        const cleanup = koffi.decode(engine.cleanup, CleanupProto);

        if (init() !== 0) {
            throw new Error(`Failed to initialize engine: ${engineName}`);
        }

        let passedConfig = config;
        if (!passedConfig) {
            passedConfig = {
                mounts: null,
                mount_count: 0,
                disable_network: 0,
                env_vars: null,
                env_count: 0,
                timeout_secs: 0,
                stdout_buffer: null,
                stdout_size: null,
                stderr_buffer: null,
                stderr_size: null,
                max_memory_mb: 0,
                max_cpu_percent: 0,
                drop_privileges: 0,
                target_uid: 0,
                target_gid: 0,
                denied_syscalls: null,
                denied_syscall_count: 0,
                seccomp_profile_path: null,
                apparmor_profile: null,
                use_pty: 0,
                max_network_mbps: 0
            };
        }

        const argc = commandArgs.length;
        const argv = commandArgs;

        const status = execute(passedConfig, argc, argv);
        cleanup();
        
        return status;
    }
}

File: bindings/sandbox.rb
require 'ffi'

# The LibMountSandbox module provides a Ruby FFI wrapper around the C libmountsandbox library.
module LibMountSandbox
  extend FFI::Library

  # Helper method to dynamically resolve the shared library path based on the host OS.
  # @param lib_path [String, nil] An optional override path to the shared library.
  # @return [String] The resolved absolute path to the shared library.
  def self.resolve_lib_path(lib_path = nil)
    return lib_path if lib_path

    lib_name = case RbConfig::CONFIG['host_os']
               when /darwin/ then 'libmountsandbox.dylib'
               when /mswin|msys|mingw|cygwin|bccwin|wince|emc/ then 'mountsandbox.dll'
               else 'libmountsandbox.so'
               end

    File.expand_path("../../build/#{lib_name}", __FILE__)
  end

  # Represents a single directory mount inside the sandbox.
  class SandboxMount < FFI::Struct
    layout :dir, :string,
           :read_only, :int
  end

  # Configuration parameters for a sandbox execution.
  class SandboxConfig < FFI::Struct
    layout :mounts, :pointer,
           :mount_count, :size_t,
           :disable_network, :int,
           :env_vars, :pointer,
           :env_count, :size_t,
           :timeout_secs, :uint,
           :stdout_buffer, :pointer,
           :stdout_size, :pointer,
           :stderr_buffer, :pointer,
           :stderr_size, :pointer,
           :max_memory_mb, :uint,
           :max_cpu_percent, :uint,
           :drop_privileges, :int,
           :target_uid, :uint,
           :target_gid, :uint,
           :denied_syscalls, :pointer,
           :denied_syscall_count, :size_t,
           :seccomp_profile_path, :string,
           :apparmor_profile, :string,
           :use_pty, :int,
           :max_network_mbps, :uint
  end

  # Defines the interface for a specific sandbox implementation.
  class SandboxEngine < FFI::Struct
    layout :engine_name, :string,
           :description, :string,
           :init, :pointer,
           :execute, :pointer,
           :execute_async, :pointer,
           :wait_process, :pointer,
           :free_process, :pointer,
           :cleanup, :pointer
  end

  # The SandboxWrapper is the main class used to execute commands via the FFI.
  class SandboxWrapper
    # Initializes the sandbox wrapper.
    # @param lib_path [String, nil] An optional path to the libmountsandbox shared library.
    def initialize(lib_path = nil)
      @lib_path = LibMountSandbox.resolve_lib_path(lib_path)
      
      # We load the library into an anonymous module to isolate state if needed,
      # but FFI::Library is typically used at the module level.
      # To match the Python API, we do it here:
      @lib = Module.new do
        extend FFI::Library
      end
      @lib.ffi_lib @lib_path
      @lib.attach_function :get_sandbox_engine, [:string, :pointer], :int
    end

    # Executes a command within the sandboxed environment.
    # @param engine_name [String] The name of the engine (e.g. 'native').
    # @param command_args [Array<String>] The command arguments to execute.
    # @param config [SandboxConfig, nil] The sandbox configuration (optional).
    # @return [Integer] The exit status of the executed command.
    def execute(engine_name, command_args, config = nil)
      engine_ptr_ptr = FFI::MemoryPointer.new(:pointer)
      res = @lib.get_sandbox_engine(engine_name, engine_ptr_ptr)

      if res != 0 || engine_ptr_ptr.null? || engine_ptr_ptr.read_pointer.null?
        raise ArgumentError, "Unknown sandbox engine: #{engine_name}"
      end

      engine = SandboxEngine.new(engine_ptr_ptr.read_pointer)

      init_func = FFI::Function.new(:int, [], engine[:init])
      execute_func = FFI::Function.new(:int, [:pointer, :int, :pointer], engine[:execute])
      cleanup_func = FFI::Function.new(:void, [], engine[:cleanup])

      if init_func.call != 0
        raise RuntimeError, "Failed to initialize engine: #{engine_name}"
      end

      argc = command_args.length
      argv = FFI::MemoryPointer.new(:pointer, argc)
      
      # Keep references to string pointers so they aren't garbage collected during the call
      str_ptrs = command_args.map { |arg| FFI::MemoryPointer.from_string(arg) }
      argv.write_array_of_pointer(str_ptrs)

      cfg = config
      if cfg.nil?
        cfg = SandboxConfig.new
        cfg.pointer.clear # Zero out the entire struct memory
      end

      status = execute_func.call(cfg.to_ptr, argc, argv)
      cleanup_func.call
      
      status
    end
  end
end
